functions/src/index.ts

```diff
@@ -52,11 +67,110 @@ export const finishGame = functions.https.onCall(async (data, context) => {
     );
   }
 
-  // Success! The game has ended, so we do an idempotent update.
-  await admin.database().ref(`games/${gameId}`).update({
-    status: "done",
-    endedAt: finalTime,
+  // The game has ended, so we attempt to do an atomic update.
+  // Safety: Events can only be appended to the game, so the final time must remain the same.
+  const { committed, snapshot }: TransactionResult = await admin
+    .database()
+    .ref(`games/${gameId}`)
+    .transaction((game) => {
+      if (game.status !== "ingame") {
+        // Someone beat us to the atomic update, so we cancel the transaction.
+        return;
+      }
+      game.status = "done";
+      game.endedAt = finalTime;
+      return game;
+    });
+
+  if (!committed) {
+    return;
+  }
+
+  /**
+   * Update ratings of players involved based on an extension of the Elo system.
+   */
+
+  // Retrieve userIds of players in the game.
+  const players: string[] = [];
+  snapshot.child("users").forEach(function (childSnap) {
+    if (childSnap.key !== null) {
+      players.push(childSnap.key);
+    }
   });
+
+  // Differentiate between single-player and multiplayer games.
+  if (players.length === 1) {
+    return;
+  }
+
+  // Add scores for players without a single set.
+  for (const player of players) {
+    scores[player] ??= 0;
+  }
+
+  // Retrieve old ratings from the database.
+  const ratings: Record<string, number> = {};
+  for (const player of players) {
+    const ratingSnap = await admin
+      .database()
+      .ref(`userStats/${player}/${gameMode}/rating`)
+      .once("value");
+    const rating = ratingSnap.exists() ? ratingSnap.val() : BASE_RATING;
+    ratings[player] = rating;
+  }
+
+  // Compute expected ratio for each player.
+  const expectedRatio: Record<string, number> = {};
```

Comment:
This name is now somewhat unrepresentative of what it contains, which may lead to confusion. Possibly rename it to just likelihood or exponentialRatings. The 'ratio' in `expectedRatio` implies that it has already been scaled with the totalLikelihood.